### PR TITLE
Fix Linux and Windows spinOnce behaviour

### PR DESCRIPTION
**Pull request by DownerCase**
This fixes two issues; one each on Windows and Linux with the behaviour of spinOnce in VTK 9.

Linux:
#5252 changed spinOnce to use ProcessEvents + sleeping to service the event queue. However, calling ProcessEvents on a vtkXRenderWindowInteractor only handles a single event and does not empty the queue so we need to repeatedly call it for smooth interaction.

Windows:
VTK 9.2 introduced this commit [VTK#8696](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8696) which prevents calling PostQuitMessage more than once if TerminateApp is called multiple times.
Problem here being the Start + Timer callback with TerminateApp method relied on this behaviour. 
When the window is in a modal event loop (most commonly due to the user resizing or moving the window) the VTK event loop does not receive the WM_QUIT message posted by TerminateApp.
Since the change in 9.2 if we miss that message the event loop will be forever stuck and spinOnce blocks forever because the event loop can't be closed.

In summary, we should call repeatedly call ProcessEvents and break out after the requested period of time or the window's Done member becomes True due to TerminateApp.

I am not throttling the rate at which ProcessEvents is called by using DO_EVERY to most closely match the behaviour of both spin(), where the window can't throttle the update rate, and the implementation of the RenderWindowInteractor::Start() methods.

I've tested this with VTK 9.1.0, 9.2.2 and the master as of 04/12/22 on Windows and Linux (Arch) and the changes work great.
If someone could verify this works on macOS then could we remove the checks for specific window interactor implementations?

### CODE REVIEW

**Comment by larshg** [comment]
Do you have a quick bullet point list how to reproduce it on windows? Ie. how do make app quit while moving the window or otherwise make the error happen?

**Comment by DownerCase** [comment]
> Do you have a quick bullet point list how to reproduce it on windows? Ie. how do make app quit while moving the window or otherwise make the error happen?

Sure. Here is a minimal example that should show the problem when using VTK 9.2 or newer

```
#include <pcl/visualization/pcl_visualizer.h>

int main (int argc, char **argv)
{
  pcl::visualization::PCLVisualizer p ("test");
  p.addCoordinateSystem (0.1);
  
  // Demonstrate usage of spinOnce()
  while (!p.wasStopped())
  {
    static int counter = 0;
    std::cout << "spinOnce was called "<<++counter<<" times.\n";
    p.spinOnce(100);  // Give the GUI 100ms to handle events, then return
  }
}
```

To trigger the issue grab the windows title bar and move the window. The console output should have stopped and if you break the program the execution is stuck in vtkWin32RenderWindowInteractor::StartEventLoop() at the while(GetMessage...) loop.
The PCL window will not close when given the 'q' and 'e' keypresses.

As mentioned this only became an issue with [VTK#8696](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8696)
On reflection this could also be solved VTK side by adding  a check for `Done` being true in the message loop. 
Though I feel it is better for PCL to repeatedly call ProcessEvents for however long is needed than the comparatively more complex callback timer + calling Start + having the timer call TerminateApp just to break out the event loop.

**Comment by DownerCase** [comment]
> I have opened an issue on vtk gitlab [here](https://gitlab.kitware.com/vtk/vtk/-/issues/18804) - since indeed the vtkXRenderWindowInteractor doesn't seem to process events in a loop until all is processed.
> 
> But windows and mac does. I'll try find time to test this, though I don't have vtk 9.2.2 installed, yet.

Thanks for raising that issue, their fix has corrected the behavior on the VTK side.
However, the call to std::this_thread::sleep_for on PCL side still needs to be removed in favor of looping on processEvents for the requested duration.

**Comment by zhangzscn** [comment]
> 
In VTK 9.2.5 still have the same problem

**Comment by mvieth** [comment]
Sorry for the delay!
I don't see a reason to use `DO_EVERY (1.0 / interactor_->GetDesiredUpdateRate (), ...)` either, so I am fine with removing it.
I found the documentation of `ProcessEvents` rather unclear, so I [asked for clarification](https://gitlab.kitware.com/vtk/vtk/-/issues/18951). Especially relevant would be whether `ProcessEvents` returns immediately if no events need processing. This seems to be the intention, so I would add a short sleep of 1ms or so after `ProcessEvents`, to avoid "busy waiting" (calling `ProcessEvents` again and again without pause even though there is no new event).

Additionally, I had the problem of `ProcessEvents` sometimes taking longer than 1 second (see the VTK issue). Creating a one-shot timer just before the loop fixed this (`interactor_->CreateOneShotTimer(time);`).

What do you think about these additions, @DownerCase ?

**Comment by DownerCase** [comment]
Your point about "busy waiting" is spot on. I recently implemented a loop similar to here and found that the program used 100% CPU to do nothing; I fixed it on Windows by adding a `WaitMessage()` call before calling `ProcessEvents()` to suspend the thread until a new message arrives. I could not immediately find the X11 equivalent. Adding a 1-10ms `sleep_for` is probably acceptable enough for this.

No issue with the one-shot timer as a fail safe for if the linked version of `ProcessEvents` starts blocking due to waiting for an event to come in.

**Comment by mvieth** [comment]
@DownerCase Great! Then please make the additions to the code in this pull request

**Review by mvieth — APPROVED** [review]
Thank you for this fix!

**Comment by larshg** [comment]
Looks good. The only thing that might be undesirable now, is that if you keep moving the camera, the application is frozen in the eventloop within VTK, which I don't think we can break out of.

Ie. even though you use spinOnce(100), it will not return after the 100ms.

This was possible on linux before at least, since it only processed one event.

Might need to clarify it in the documentation.

**Comment by mvieth** [comment]
@larshg You mean because there is no guarantee that `spinOnce` will return exactly after the given time, instead it will return when the time is done _and_ there are no more events to process?
Let's maybe add that to the doc in a separate pull request and get this one merged soon?

**Comment by larshg** [comment]
> @larshg You mean because there is no guarantee that `spinOnce` will return exactly after the given time, instead it will return when the time is done _and_ there are no more events to process? Let's maybe add that to the doc in a separate pull request and get this one merged soon?

Yes, currently I would expect, given the documentation that the function will return after the time supplied. But it won't, if the event queue is flooded. Ie. by panning the camera contineously.

I think it has been like this for a long time, so surely this is an improvement, but the documentation would need an update.

**Review by larshg — APPROVED** [review]
